airbyte-integrations/connectors/source-oracle-v2/src/cdk/main/kotlin/io/airbyte/cdk/AirbyteConnectorRunner.kt

```diff
@@ -0,0 +1,182 @@
+import picocli.CommandLine.Model.UsageMessageSpec
+
+/** Source connector entry point. */
+class AirbyteSourceRunner(args: Array<out String>) :
+    AirbyteConnectorRunner(ConnectorType.SOURCE, args) {
+
+    companion object {
+        @JvmStatic
+        fun run(vararg args: String) {
+            AirbyteSourceRunner(args).run<AirbyteConnectorRunnable>()
+        }
+    }
+}
+
+/** Destination connector entry point. */
+class AirbyteDestinationRunner(args: Array<out String>) :
+    AirbyteConnectorRunner(ConnectorType.DESTINATION, args) {
+
+    companion object {
+        @JvmStatic
+        fun run(vararg args: String) {
+            AirbyteDestinationRunner(args).run<AirbyteConnectorRunnable>()
+        }
+    }
+}
+
+/**
+ * Replacement for the Micronaut CLI application runner that configures the CLI components and adds
+ * the custom property source used to turn the arguments into configuration properties.
+ */
+sealed class AirbyteConnectorRunner(
+    val connectorType: ConnectorType,
+    val args: Array<out String>,
+) {
+
+    val envs: Array<String> = arrayOf(Environment.CLI, connectorType.name.lowercase())
+
+    inline fun <reified R : Runnable> run() {
+        val micronautCommandLine: MicronautCommandLine = MicronautCommandLine.parse(*args)
+        val configPropertySource = ConnectorCommandLinePropertySource(micronautCommandLine)
+        val commandLinePropertySource = CommandLinePropertySource(micronautCommandLine)
+        val ctx: ApplicationContext =
+            ApplicationContext.builder(R::class.java, *envs)
+                .propertySources(configPropertySource, commandLinePropertySource)
+                .start()
+        val isTest: Boolean = ctx.environment.activeNames.contains(Environment.TEST)
+        val picocliFactory: CommandLine.IFactory = MicronautFactory(ctx)
+        val picocliCommandLine: CommandLine =
+            PicocliCommandLineFactory(connectorType, isTest)
+                .build<AirbyteConnectorRunnable>(picocliFactory)
+        try {
+            picocliCommandLine.execute(*args)
+        } catch (e: Throwable) {
+            // Many of the exceptions thrown are nested inside layers of RuntimeExceptions. An
+            // attempt is made to find the root exception that corresponds to a configuration
+            // error. If that does not exist, we just return the original exception.
+            ApmTraceUtils.addExceptionToTrace(e)
+            val rootThrowable = ConnectorExceptionUtil.getRootConfigError(Exception(e))
+            val displayMessage = ConnectorExceptionUtil.getDisplayMessage(rootThrowable)
+            // If the connector throws a config error, a trace message with the relevant
+            // message should be surfaced.
+            if (ConnectorExceptionUtil.isConfigError(rootThrowable)) {
+                AirbyteTraceMessageUtility.emitConfigErrorTrace(e, displayMessage)
+            }
+        }
+    }
+}
+
+enum class ConnectorType {
```

Comment:
looks like what you really want is a sealed class if you want inheritance + enum, no?

Reply:
I think I get it. Yep!